### PR TITLE
graphql: add support for tx types and tx access lists

### LINKED ISSUE #22487: GraphQL schema is not EIP-2718/EIP-2930 aware
#### System information

Latest master, last relevant commit https://github.com/ethereum/go-ethereum/commit/39b3b8ffb44983a260031d5077226d952ddfb9bf

#### Expected behaviour

The GraphQL schema for `Transaction` should have `Type` and `AccessList` fields:

https://github.com/ethereum/go-ethereum/blob/master/graphql/schema.go#L73

#### Actual behaviour

The schema was not updated for EIP-2718/-2930.  With EIP-1559 on the horizon it might make sense long-term to refactor `Transaction` into a [union type](https://graphql.org/learn/schema/#union-types), something like `union Transaction = LegacyTransaction | AccessListTransaction | BaseFeeTransaction`.

#### Steps to reproduce the behaviour

As reported on go-ethereum discord the access list and type fields aren't queryable via GraphQL, obviously because they are not yet included in the schema.

### PR DESCRIPTION
**Pull request by AmitBRD**
fixes #22487  this PR adds the ability to query the type and AccessTuple from graphql endpoint

### CODE REVIEW

File: graphql/graphql.go [hunk 1]
```diff
@@ -151,6 +151,20 @@ func (l *Log) Data(ctx context.Context) hexutil.Bytes {
 	return l.log.Data
 }
 
+// Access Tuple represents EIP-2930
+type AccessTuple struct {
+	address     common.Address
+	storageKeys *[]common.Hash
+}
```
Comment (holiman) [inline graphql/graphql.go:158]: We already have 
```
// AccessList is an EIP-2930 access list.
type AccessList []AccessTuple

// AccessTuple is the element type of an access list.
type AccessTuple struct {
	Address     common.Address `json:"address"        gencodec:"required"`
	StorageKeys []common.Hash  `json:"storageKeys"    gencodec:"required"`
}
```
In `core/types`. Is there any reason not to just use those?

File: graphql/graphql.go [hunk 2]
```diff
@@ -151,6 +151,20 @@ func (l *Log) Data(ctx context.Context) hexutil.Bytes {
 	return l.log.Data
 }
 
+// Access Tuple represents EIP-2930
```
Comment (holiman) [inline graphql/graphql.go:154]: docstring error, needs to start with `// AccessTuple`

File: graphql/graphql.go [hunk 3]
```diff
@@ -342,6 +356,31 @@ func (t *Transaction) Logs(ctx context.Context) (*[]*Log, error) {
...
+	txType := int32(tx.Type())
+	return &txType, nil
+}
+
+func (t *Transaction) AccessList(ctx context.Context) (*[]*AccessTuple, error) {
+	tx, err := t.resolve(ctx)
+	if err != nil || tx == nil {
+		return nil, err
+	}
+	accessList := tx.AccessList()
```
Comment (holiman) [inline graphql/graphql.go:373]: Why not just return the `tx.AccessList` ?

**Review by holiman — COMMENTED** [review]
Thanks for submitting the PR!

In addition to the inlined comments, we would need some testcases for this, alongside the existing graphql tests.

File: graphql/graphql.go (hunk 1 above)
Comment (AmitBRD) [reply graphql/graphql.go:158]: Ill add testcases to graphql_test.go for type and accessList properties retrieval.  My initial line of thinking was separating the graphql and core object models, but I directly use the tuples if that is preferred.

File: graphql/graphql.go (hunk 3 above)
Comment (fjl) [reply graphql/graphql.go:373]: It's not possible to just return it because GraphQL requires wrapping API objects in a certain way.

**Review by fjl — APPROVED** [review]
Tested this using 

```
{
  block (number:8314243) {
    number
    transactions {
      type
      accessList {
        address
        storageKeys
      }
    }
  }
}
```

and it works on Rinkeby.

**Pull request merged**